packages/core/js-client/src/jsPeer/FluencePeer.ts

```diff
@@ -303,47 +325,193 @@ export abstract class FluencePeer {
+                            // MUST happen sequentially (in a critical section).
+                            // Otherwise the race might occur corrupting the prevData
+
+                            log_particle.debug('id %s. sending particle to interpreter', item.particle.id);
+                            log_particle.trace('id %s. prevData: %s', item.particle.id, this.decodeAvmData(prevData));
+
+                            const args = serializeAvmArgs(
+                                {
+                                    initPeerId: item.particle.initPeerId,
+                                    currentPeerId: this.keyPair.getPeerId(),
+                                    timestamp: item.particle.timestamp,
+                                    ttl: item.particle.ttl,
+                                    keyFormat: KeyPairFormat.Ed25519,
+                                    particleId: item.particle.id,
+                                    secretKeyBytes: this.keyPair.toEd25519PrivateKey(),
+                                },
+                                item.particle.script,
+                                prevData,
+                                item.particle.data,
+                                item.callResults,
+                            );
+
+                            let avmCallResult: InterpreterResult | Error;
+                            try {
+                                const res = await this.marineHost.callService('avm', 'invoke', args, defaultCallParameters);
+                                avmCallResult = deserializeAvmResult(res);
+                            } catch (e) {
+                                avmCallResult = e instanceof Error ? e : new Error(String(e));
+                            }
+
+                            if (!(avmCallResult instanceof Error) && avmCallResult.retCode === 0) {
+                                const newData = Buffer.from(avmCallResult.data);
+                                prevData = newData;
+                            }
+
+                            return {
+                                ...item,
+                                result: avmCallResult,
+                            };
+                        }),
+                        filter((item): item is NonNullable<typeof item>  => item !== null),
+                        filterExpiredParticles<ParticleQueueItem & {result: Error | InterpreterResult }>(this._expireParticle.bind(this)),
+                        concatMap(async (item) => {
+                            // If peer was stopped, do not proceed further
+                            if (item === null || !this.isInitialized) {
+                                return;
+                            }
+
+                            // Do not proceed further if the particle is expired
+                            if (hasExpired(item.particle)) {
+                                return;
+                            }
```

Comment:
are these two checks still needed if you already filtered above? I never used rx-js enough to understand everything, so probably my concern is invalid

Reply:
You understand it correctly. Will fix